### DISCUSSION TITLE
Database Connection Via SSH Tunneling?

### DISCUSSION THREAD

**Discussion by hatchli** [prisma/prisma#2098] [Q&A] [answered]
This doesn't seem to be documented at all. Is it even possible to connect the prisma client via an SSH tunnel? I can successfully connect via Workbench, but not sure how to build the connection string as described in the docs.

I have:
SSH Hostname
SSH Username
SSH Password
MySQL Hostname (localhost)
MySQL Port
MySQL username
MySQL password

**Comment by nikolasburk** [comment]
Hey @hatchli 👋 have you seen this [page](https://www.prisma.io/docs/reference/database-connectors/mysql#configuring-an-ssl-connection) in the docs? If that doesn't help, can you share the open questions you have? 🙂

**Reply by hatchli** [reply]
Hey @nikolasburk! I have reviewed that page and tried to work from it. My understanding is that SSL is different from SSH tunneling. I'll provide a screenshot of how I'm accessing the db via Workbench if that clarifies things. Potentially that SSL section you referenced would work, I just don't understand which parameter maps to which! 

<img width="658" alt="mysqlsettings" src="https://user-images.githubusercontent.com/57643179/78575355-3b14c380-77f9-11ea-8d50-db0c2cb54850.png">
While I don't use an SSH Key (or if I do I'm not aware) I do use a SSH user password and a MySQL user password.

So using the page you referenced, I couldn't figure out how I would build the connection string. Any idea?

**Comment by pimeys** [comment, marked as answer]
So basically it works like this. If you can connect to the database machine through ssh, do so first:

```sh
ssh -L 3306:localhost:3306 database-machine.org
```

This makes an assumption your database listens on localhost on your server and listens port 3306. Also makes an assumption you have an free port 3306 on your machine.

So, how it works: the first 3306 on the string is the port you open on localhost, the middle part `localhost` means the hostname the MySQL server is listening to (can also be `127.0.0.1` or `0.0.0.0`) and the last number is the port the server is listening to.

When you get this running, from the same machine you can modify your connection string to:

```
mysql://user:pass@localhost:3306/database
```

And it should work. I use this method all the time for development.

**Reply by skylinezum** [reply]
How would it work if the database is in a private subnet and I have an instance on a public subnet which I ssh over to access the database?

**Reply by ryands17** [reply]
@skylinezum 👋 
It would work fine. You will get access to your database on `localhost`, so something like:

```
ssh -N -L 5432:hostname:5432 instance@ipAddress -i keyName.pem
```

**Reply by depsimon** [reply]
Thanks for the explanation. 

How can we use that in a Node.js project? (using Next.js on Vercel)

Is that recommended or are there any alternatives?

**Reply by janacm** [reply]
For those trying to connect to Google Cloud SQL running on private network, from NodeJS/Prisma, you first need to start an SSH tunnel (this is the `-N` flag) before trying to connect.

1. First open the tunnel like this:
```
ssh -N -L 63333:<gcp db private ip>:5432 <db user>@<gcp vpc host> -i ~/.ssh/<your ssh private key>
```
Your database url should look like: 
```
DATABASE_URL=postgres://postgres:<your db pw>@127.0.0.1:63333/<your db name>
```

Note: if your app is also running in GCP, then you should have access to the GCP db without having to tunnel.

**Reply by ryancwalsh** [reply]
I appreciate your comments, but I'm still confused.

Can you please see my question at https://stackoverflow.com/questions/75502931/how-to-securely-connect-to-a-database-in-next-js-app-using-prisma-and-mysql-de?

Thanks!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
